kmk/extensions/ble_split.py

```diff
@@ -0,0 +1,147 @@
+
+    def _to_dict(self):
+        return f'BLE_Split( _ble={self._ble} _ble_last_scan={self._ble_last_scan} _is_target={self._is_target} _uart_buffer={self._uart_buffer} _split_flip={self.split_flip} _split_side={self.split_side} )'
+
+    def during_bootup(self, keyboard):
+        self._is_target = bool(self.split_side == 'Left')
+
+        if self.split_flip and not self._is_target:
+            keyboard.col_pins = list(reversed(keyboard.col_pins))
+
+        self.split_offset = len(keyboard.col_pins)
+
+        # Attempt to sanely guess a coord_mapping if one is not provided.
+        if not keyboard.coord_mapping:
+            keyboard.coord_mapping = []
+
+            rows_to_calc = len(keyboard.row_pins) * 2
+            cols_to_calc = len(keyboard.col_pins) * 2
+
+            for ridx in range(rows_to_calc):
+                for cidx in range(cols_to_calc):
+                    keyboard.coord_mapping.append(intify_coordinate(ridx, cidx))
+
+    def before_matrix_scan(self, keyboard_state):
+        self.check_all_connections()
+        return self._receive()
+
+    def after_matrix_scan(self, keyboard_state, matrix_update):
+        if matrix_update:
+            matrix_update = self._send(matrix_update)
+            return matrix_update
+
+    def check_all_connections(self):
+        self._connection_count = len(self._ble.connections)
+        if self._is_target and self._connection_count < 2:
+            self.target_advertise()
+        elif not self._is_target and self._connection_count < 1:
+            self.initiator_scan()
+
+    def initiator_scan(self):
+        self._uart = None
+        self._uart_connection = None
+        # See if any existing connections are providing UARTService.
+        self._connection_count = len(self._ble.connections)
+        if self._connection_count > 0 and not self._uart:
+            for connection in self._ble.connections:
+                if UARTService in connection:
+                    self._uart_connection = connection
+                    self._uart = self._uart_connection[UARTService]
+                    break
+
+        if not self._uart:
+            print('Scanning')
+            self._ble.stop_scan()
+            for adv in self._ble.start_scan(ProvideServicesAdvertisement, timeout=20):
+                print('Scanning')
+                if UARTService in adv.services:
+                    self._uart_connection = self._ble.connect(adv)
+                    self._uart = self._uart_connection[UARTService]
+                    self._ble.stop_scan()
+                    print('Scan complete')
+                    break
+        self._ble.stop_scan()
+        return
+
+    def target_advertise(self):
+        self._ble.stop_advertising()
+        print('Advertising')
+        # Uart must not change on this connection if reconnecting
+        if not self._uart:
+            self._uart = UARTService()
+        advertisement = ProvideServicesAdvertisement(self._uart)
```

Comment:
Willing to accept commits on this part. I know very little about BLE in general. It would need to be configured in some sane way that the keyboard knew on both sides what to look for as well as the only thing that will look for this, is the right half. Maybe uuid?

Reply:
I've actually done this though it doesn't have an effect. It changes it's name now based on it's volume name, and can even get which side it is by the last letter "L" or "R".

Reply:
I'll address this in a follow up PR. This one is too big.